File: articles/equilibrium-dynamics-reversible-systems/julia/equilibrium_kernel.jl
# Chemical equilibrium numerical kernel in Julia.

const R = 8.314462618

function delta_g_from_qk(Q, K, T)
    return R * T * log(Q / K) / 1000.0
end

function solve_isomerization(K, total)
    Aeq = total / (1.0 + K)
    Beq = total - Aeq
    return Aeq, Beq
end

function reversible_step(A, B, kf, kr, dt)
    net = kf * A - kr * B
    return max(A - net * dt, 0.0), max(B + net * dt, 0.0)
end

Aeq, Beq = solve_isomerization(4.0, 1.0)
dg = delta_g_from_qk(0.5, 4.0, 298.15)
A1, B1 = reversible_step(1.0, 0.0, 0.20, 0.05, 0.25)

println("A_eq=", round(Aeq, digits=6))
println("B_eq=", round(Beq, digits=6))
println("delta_g_kj_mol=", round(dg, digits=6))
println("A_after_step=", round(A1, digits=6))
println("B_after_step=", round(B1, digits=6))

File: articles/equilibrium-dynamics-reversible-systems/fortran/equilibrium_kernel.f90
program equilibrium_kernel
  implicit none

  real(8), parameter :: R = 8.314462618d0
  real(8) :: K, total, Aeq, Beq, Q, T, delta_g

  K = 4.0d0
  total = 1.0d0
  Aeq = total / (1.0d0 + K)
  Beq = total - Aeq

  Q = 0.5d0
  T = 298.15d0
  delta_g = R * T * log(Q / K) / 1000.0d0

  print *, "A_eq=", Aeq
  print *, "B_eq=", Beq
  print *, "delta_g_kj_mol=", delta_g

end program equilibrium_kernel
